### ISSUE TITLE
problem with .onLoad in RStudio?

### ISSUE THREAD

**Issue by davharris** [ropensci/ecoretriever#49]
I just filed [a bug report with Rstudio](https://support.rstudio.com/hc/communities/public/questions/203904787-Unexpected-exception-value-type-is-5-not-0) about this, but I thought it made sense to cross-post.

*Apparently*, RStudio is using .onLoad when I open certain scripts or projects that use this package.  And when `retriever update` is finished, it doesn't know what to do with the package output and crashes.

Can someone try reproducing, using the steps I posted on RStudio's support page to confirm that I'm not completely off-base?  It might only break on [the Preview version of RStudio](http://www.rstudio.com/products/rstudio/download/preview/).

Thanks!

**Comment by dmcglinn** [comment]
I just posted to the Rstudio thread that I wasn't able to reproduce this behavior in windows. If you are correct about why your system is crashing then it should be due to this line of code `system('retriever update', ignore.stdout=FALSE, ignore.stderr=TRUE)`. Run that and see if it causes the same pattern of behavior in Rstudio. There is no need to have `ecoretriever` loaded to run this bit of code.

**Comment by kevinushey** [comment]
This failure is now worked around in RStudio, but you should understand the cause.

RStudio attempts to provide completions for packages discovered in source documents (even without loading them directly in the current R session), by essentially doing the following:

1. Starting a new R process,
2. Loading the package (attempting to suppress as much output as possible),
3. Passing back a JSON-encoded object back on stdout to the other R process.

The issue with `ecoretriever` is that the 'updating scripts' lines normally printed on stdout 'leak' into the output that RStudio wants to pass back, e.g.:

```
Downloading script: MammalSuperTree.py
0.0-MammalSuperTree.py 100.0-MammalSuperTree.py Downloading script: EA_capellini2010.script
0.0-EA_capellini2010.script 100.0-EA_capellini2010.script Downloading script: EA_avianbodysize2007.script
0.0-EA_avianbodysize2007.script 100.0-EA_avianbodysize2007.script Downloading script: EA_mom2003.script
0.0-EA_mom2003.script 100.0-EA_mom2003.script Downloading script: EA_pantheria.script
0.0-EA_pantheria.script 100.0-EA_pantheria.script Downloading script: bbs.py
0.0-bbs.py 80.7-bbs.py 100.0-bbs.py Downloading script: EA_del_moral_2010.script
0.0-EA_del_moral_2010.script 100.0-EA_del_moral_2010.script Downloading script: EA_mcglinn2010.script
0.0-EA_mcglinn2010.script 100.0-EA_mcglinn2010.script Downloading script: EA_barnes2008.script
0.0-EA_barnes2008.script 100.0-EA_barnes2008.script Downloading script: EA_ramesh2010.script
```

RStudio ends up trying to parse that as JSON and explodes. Note that e.g.

    capture.output(library(ecoretriever))

does not work, as the output is sent directly to stdout and R doesn't get to intern / handle it (this is how we normally protect RStudio from 'noisy' package startup)

In my opinion, `ecoretriever` could do a couple things to be a 'better citizen' here:

1. Don't write to stdout in such a way that `capture.output` cannot capture it,
2. Don't automatically update package scripts on load; leave that for a separate function (perhaps let the package startup message ask the users to call that function)

**Comment by dmcglinn** [comment]
Thanks @kevinushey for the very helpful insight. @ethanwhite any thoughts on this? How would you feel about turning off the update scripts on package load feature? Additionally I should be able to change how the script updating information is printed to screen so that it is less prone to interference with other processes.

**Comment by ethanwhite** [comment]
I agree that we should turn off the update scripts on loading the package and replace this with clear documentation. I like @kevinushey's suggestion of including those instructions in the package startup message.

Not being an R person I don't know what "don't write to stdout in such a way that `capture.output` cannot capture it" entails, but I agree that it's something we should do.

**Comment by gavinsimpson** [comment]
@dmcglinn I agree with this; the package should be loadable without side effects or waiting for updates. You'll need some UI function to allow the user to initiate those updates, say `refresh()`? `update()` has different connotations in R - it's used to update fitted models - so you probably want to avoid that.

**Comment by kevinushey** [comment]
I would say something like `ecoretriever::initialize(update = TRUE)` would be nice.

**Comment by gavinsimpson** [comment]
@kevinushey to me `initialize()` sounds more like starting something and `ecoretrier` (the pkg) is already started and even when interfacing with the ecoretriever binary it's not like that is left in some running state that you might want to initialise.

**Comment by gavinsimpson** [comment]
@dmcglinn @ethanwhite I think @kevinushey means that `retriever update` is writing to stdout, which gets injected into the R console when you load the package but `capture.output()` can't see that injection. I'm not familiar with system calls, but `system2` seems more recommended these days and you can have `system2()` capture stdout and stderr and return these as character vectors, which you would then write as output from the R side using say `message()` or `writeLines()` even. That way I think RStudio would be able to capture this output.

Apologies if I've got this wrong or misunderstood Kevin's comment.

**Comment by dmcglinn** [comment]
Take a look at my PR in #51 and let me know what you think.

**Comment by ethanwhite** [comment]
Should be fixed by #51.
